CONTRIBUTING.md

```diff
@@ -47,44 +48,47 @@ the expected behavior that doesn't occur.
 
 To suggest an enhancement, open an
 [issue](https://github.com/nginx/unit/issues/new) on GitHub with the label
-`enhancement`.  Please do this before implementing a new feature to discuss the
-feature first.
+`enhancement`.  Please do this before implementing a new feature to discuss
+the feature first.
 
 
 ### Open a Pull Request
 
-Before submitting a PR, please read the NGINX Unit code guidelines to know more
-about coding conventions and benchmarks.  Fork the repo, create a branch, and
-submit a PR when your changes are tested and ready for review.  Again, if you'd
-like to implement a new feature, please consider creating a feature request
-issue first to start a discussion about the feature.
+Before submitting a PR, please read the NGINX Unit code guidelines to know
+more about coding conventions and benchmarks.  Fork the repo, create a branch,
+and submit a PR when your changes are tested and ready for review.  Again, if
+you'd like to implement a new feature, please consider creating a feature
+request issue first to start a discussion about the feature.
 
 
 ## Git Style Guide
 
-- Keep a clean, concise and meaningful `git commit` history on your branch,
-  rebasing locally and squashing before submitting a PR
+- Create atomic commits.  A commit should do just one thing, I.e you
+  shouldn't mix whitespace changes with actual code changes.  Do the
+  whitespace changes in a separate commit first.
```

Comment:
I don't think the order is important.
```suggestion
  whitespace changes in a separate commit.
```

Reply:
I disagree.

In general you should do any cleanup patches first, e.g as part of preparatory patches for a change.

It's the same as fixing a bunch of coding style issues on some code _before_ you make actual changes to _that_ code.

Reply:
I still think cleanup can come as a later step as long as that's also included. I don't think being prescriptive about having to do cleanup before anything happens is useful. As long as cleanup happens, why should we care in what order that happened?

Sure, individual preferences exist, but those are individual, and we shouldn't push those on others.

Reply:
Except for when you're the one doing the review, I know I'm not the only one that would ask for it to be done that way...

The ordering is somewhat logical. You do preparatory work before _actual_ work.

Let's assume the whitespace changes are trivial and uncontentious, if you _did_ them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

It may also help with things like git-blame(1).

Reply:
`whitespace` changes was just an example, replace that with `refactor`, then tell me that shouldn't come first...

Reply:
It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.

Others might want to add the new / changed functionality first, and then tidy it up.

Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

I think we should be okay with all kinds of workflows and not prescribe a rigid order.

Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

Reply:
> if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

Reply:
> It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.
> 
> Others might want to add the new / changed functionality first, and then tidy it up.
> 
> Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

If you do things right there shouldn't be any _need_ to _tidy_ things up afterwards. I.e. this is bad form

```
Fixup previous commit
Introduce new feature
```

All it does is add extra noise.

If you do your refactor/cleanup first then it is often the case that your subsequent commits are _simpler_ and _more concise_.

You want commits to be as small and concise as possible. This really helps with reviewing for one thing.

> I think we should be okay with all kinds of workflows and not prescribe a rigid order.
> 
> Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

Not really.

At the end of day it's up to the project maintainers how things are done. That is all this commit is doing, updating the contributing document to reflect how things are _currently_ done in Unit.

If you disagree with how the Unit project is currently run then that is a separate conversation.

> > if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.
> 
> I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace

Only from the point of the rebase...

> change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

In Unit (like in many other projects) the `prefix` has a specific purpose and saying what _kind_ of commit it is isn't it... also I hope you didn't just say, "_no need to review whitespace changes_"... because they should be reviewed along with everything else.

[ Also just because a patch might be labelled as `whitespace` doesn't mean there won't be _other_ changes in there! ]